src/pathfinding.cpp

```diff
@@ -0,0 +1,352 @@
+#include "coordinates.h"
+#include "debug.h"
+#include "enums.h"
+#include "game.h"
+#include "map.h"
+#include "map_iterator.h"
+#include <algorithm>
+#include <queue>
+
+#include "messages.h"
+
+enum astar_state {
+ ASL_NONE,
+ ASL_OPEN,
+ ASL_CLOSED
+};
+
+// Turns two indexed to a 2D array into an index to equivalent 1D array
+constexpr int flat_index( const int x, const int y ) {
+    return (x * MAPSIZE * SEEY) + y;
+};
+
+struct pair_greater_cmp
+{
+    bool operator()( const std::pair<int, tripoint> &a, const std::pair<int, tripoint> &b)
+    {
+        return a.first > b.first;
+    }
+};
+
+// Flattened 2D array representing a single z-level worth of pathfinding data
+struct path_data_layer
+{
+    // State is accessed way more often than all other values here
+    std::array< astar_state, SEEX * MAPSIZE * SEEY * MAPSIZE > state;
+    std::array< int, SEEX * MAPSIZE * SEEY * MAPSIZE > score;
+    std::array< int, SEEX * MAPSIZE * SEEY * MAPSIZE > gscore;
+    std::array< tripoint, SEEX * MAPSIZE * SEEY * MAPSIZE > parent;
```

Comment:
Is there a specific reason you use a 1-D array to store 2-D data? Why not use 2 arrays and let the compiler figure out the index (instead of having a `flat_index` function)?